src/controller/java/AndroidCallbacks.cpp

```diff
@@ -281,6 +283,34 @@ void ReportCallback::OnAttributeData(const app::ConcreteDataAttributePath & aPat
     env->CallVoidMethod(mNodeStateObj, addAttributeMethod, static_cast<jint>(aPath.mEndpointId),
                         static_cast<jlong>(aPath.mClusterId), static_cast<jlong>(aPath.mAttributeId), attributeStateObj);
     VerifyOrReturnError(!env->ExceptionCheck(), env->ExceptionDescribe());
+
+    UpdateClusterDataVersion();
+}
+
+void ReportCallback::UpdateClusterDataVersion()
+{
+    JNIEnv * env = JniReferences::GetInstance().GetEnvForCurrentThread();
+    chip::app::ConcreteClusterPath lastConcreteClusterPath;
+    if (mClusterCacheAdapter.GetLastReportDataPath(lastConcreteClusterPath) == CHIP_NO_ERROR &&
+        lastConcreteClusterPath.IsValidConcreteClusterPath())
+    {
+        chip::Optional<chip::DataVersion> committedDataVersion;
+        if (mClusterCacheAdapter.GetVersion(lastConcreteClusterPath, committedDataVersion) == CHIP_NO_ERROR)
+        {
+            if (committedDataVersion.HasValue())
+            {
+                // SetDataVersion to NodeState
+                jmethodID setDataVersionMethod;
+                CHIP_ERROR err =
+                    JniReferences::GetInstance().FindMethod(env, mNodeStateObj, "setDataVersion", "(IJI)V", &setDataVersionMethod);
+                VerifyOrReturn(err == CHIP_NO_ERROR, ChipLogError(Controller, "Could not find setDataVersion method"));
+                env->CallVoidMethod(mNodeStateObj, setDataVersionMethod, static_cast<jint>(lastConcreteClusterPath.mEndpointId),
+                                    static_cast<jlong>(lastConcreteClusterPath.mClusterId),
+                                    static_cast<jint>(committedDataVersion.Value()));
+                VerifyOrReturnError(!env->ExceptionCheck(), env->ExceptionDescribe());
```

Comment:
Should this be a verify-or-return? Since this method is void.

Maybe consider not using the return macros at all as the if() would be readable as long as we decrease nesting.

Reply:
updated across this file, thanks